Content.Server/Disease/DiseaseSystem.cs

```diff
@@ -0,0 +1,420 @@
+        [Dependency] private readonly DoAfterSystem _doAfterSystem = default!;
+        [Dependency] private readonly PopupSystem _popupSystem = default!;
+        [Dependency] private readonly EntityLookupSystem _lookup = default!;
+        [Dependency] private readonly SharedInteractionSystem _interactionSystem = default!;
+        [Dependency] private readonly InventorySystem _inventorySystem = default!;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+            SubscribeLocalEvent<DiseaseCarrierComponent, CureDiseaseAttemptEvent>(OnTryCureDisease);
+            SubscribeLocalEvent<DiseasedComponent, InteractHandEvent>(OnInteractDiseasedHand);
+            SubscribeLocalEvent<DiseasedComponent, InteractUsingEvent>(OnInteractDiseasedUsing);
+            SubscribeLocalEvent<DiseaseProtectionComponent, GotEquippedEvent>(OnEquipped);
+            SubscribeLocalEvent<DiseaseProtectionComponent, GotUnequippedEvent>(OnUnequipped);
+            SubscribeLocalEvent<DiseaseVaccineComponent, AfterInteractEvent>(OnAfterInteract);
+            SubscribeLocalEvent<DiseaseVaccineComponent, ExaminedEvent>(OnExamined);
+            /// Private events stuff
+            SubscribeLocalEvent<TargetVaxxSuccessfulEvent>(OnTargetVaxxSuccessful);
+            SubscribeLocalEvent<VaxxCancelledEvent>(OnVaxxCancelled);
+        }
+
+        private Queue<EntityUid> AddQueue = new();
+        private Queue<(DiseaseCarrierComponent, DiseasePrototype)> CureQueue = new();
+
+        /// <summary>
+        /// First, adds or removes diseased component from the queues and clears them.
+        /// Then, iterates over every diseased component to check for their effects
+        /// and cures
+        /// </summary>
+        public override void Update(float frameTime)
+        {
+            base.Update(frameTime);
+            foreach (var entity in AddQueue)
+                EnsureComp<DiseasedComponent>(entity);
+            AddQueue.Clear();
+
+            foreach (var tuple in CureQueue)
+            {
+                if (tuple.Item1.Diseases.Count == 1) //This is reliable unlike testing Count == 0 right after removal for reasons I don't quite get
+                    RemComp<DiseasedComponent>(tuple.Item1.Owner);
+                tuple.Item1.PastDiseases.Add(tuple.Item2);
+                tuple.Item1.Diseases.Remove(tuple.Item2);
+            }
+            CureQueue.Clear();
+
+            foreach (var (diseasedComp, carrierComp, mobState) in EntityQuery<DiseasedComponent, DiseaseCarrierComponent, MobStateComponent>(false))
+            {
+                if (mobState.IsDead())
+                {
+                    if (_random.Prob(0.005f * frameTime)) //Mean time to remove is 200 seconds per disease
+                        CureDisease(carrierComp, _random.Pick(carrierComp.Diseases));
+                    continue;
+                }
+
+                foreach(var disease in carrierComp.Diseases)
```

Comment:
this will only run on things with `MobState` meaning things without it cant ever be diseased, is that intended? if not, this `IsDead` check becomes optional and should probably be moved after a new check for whether it actually has any diseases

Reply:
This is intentional, if there's ever a component for objects it'll be a separate component and only use events